core/src/main/java/org/apache/spark/memory/TaskMemoryManager.java

```diff
@@ -507,4 +525,19 @@ public long getMemoryConsumptionForThisTask() {
   public MemoryMode getTungstenMemoryMode() {
     return tungstenMemoryMode;
   }
+
+  /**
+   * Returns peak task-level off-heap memory usage in bytes.
+   *
+   */
+  public long getPeakOnHeapExecutionMemory() {
+    return peakOnHeapMemory;
+  }
```

Comment:
Brainstorming: 

Do we need to worry about synchronization for the field accesses here (and in the other getter)? Right now the writes to these fields are guarded by `synchronized (this)` but the reads are unsynchronized.

If we did synchronize these, would we have to worry about deadlocks? I don't think so, because TaskMemoryManager is internal to Spark and right now the only call to these getters occurs once at the end of the task during final metrics emission (and occurs while not holding other TMM / MM locks). Also, TaskMemoryManager is not a DeveloperAPI thus we wouldn't expect third-party or user code to be calling this during task execution.

If we want to sidestep any potential deadlock concerns, though, then we could make both longs `volatile`, similar to the existing `private volatile long acquiredButNotUsed` (see https://github.com/apache/spark/pull/12681).

Reply:
So far I think it's safe, as you mentioned, only invoked at the end of task. But no harm to add `volatile`, in case we might have race condition in the future if we add other use case.